### PR TITLE
Implement try finally to predestroy / release managed object

### LINKED ISSUE #11453: Potential leak caused by JSTL <c:redirect> tag
The JSP engine generates code wrapped with the following pattern for tag constructs:

    _jspx_iaHelper.doPostConstruct(_jspx_xxxxxx_0);
    _jspx_iaHelper.addTagHandlerToCdiMap(_jspx_xxxxx_0, _jspx_xxxxx_0_mo);

    tag specific code here

    _jspx_iaHelper.doPreDestroy(_jspx_xxxxx_0);
    _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_xxxxx_0);

The issue is that JSP for a <c:redirect> generates a return, skipping the doPreDestroy and cleanUpTagHandlerFromCdiMap functions needed for cleanup.  The effect is a memory leak, which if used heavily enough has the potential for OOM.

### PR DESCRIPTION
**Pull request by volosied**
fixes #11453

Implements try - finally with the Tag methods to always clean up resources. For example:

```
  private boolean _jspx_meth_c_out_23(javax.servlet.jsp.tagext.JspTag _jspx_th_c_forTokens_11, PageContext pageContext, int[] _jspx_push_body_count_c_forTokens_11)
     throws Throwable {
    JspWriter out = pageContext.getOut();
    com.ibm.ws.managedobject.ManagedObject _jspx_th_c_out_23_mo = _jspx_iaHelper.inject(org.apache.taglibs.standard.tag.rt.core.OutTag.class);
    org.apache.taglibs.standard.tag.rt.core.OutTag _jspx_th_c_out_23 = (org.apache.taglibs.standard.tag.rt.core.OutTag)_jspx_th_c_out_23_mo.getObject();
    try {
       ...
      int _jspx_eval_c_out_23 = _jspx_th_c_out_23.doStartTag();
      if (_jspx_th_c_out_23.doEndTag() == javax.servlet.jsp.tagext.Tag.SKIP_PAGE) {
        return true;
      }
      return false;
    } finally {
      _jsp_cleanUpTag(_jspx_th_c_out_23, null);
    }
  }
```

For the jsp service method (and JSP Fragments invoke methods), I created a list which has tags added and removed. In the finally method, if any tags are left over, they are cleaned up.

```
java.util.ArrayList _jspTagList = new java.util.ArrayList();

try { 
...
 _jsp_tagPostConstruct(_jspx_th_c_catch_0, _jspTagList, _jspx_th_c_catch_0_mo);
...
_jsp_cleanUpTag(_jspx_th_c_catch_0, _jspTagList);
...
    } finally {
      _jsp_cleanUpTagArrayList(_jspTagList);
      _jspxFactory.releasePageContext(pageContext);
    }
```

This fix handles various scenarios (Tag Pooling & Resource Injection) to avoid regressions. 

- Tag Pooling Disabled +  Resource Injection Disabled
- Tag Pooling Enabled  +  Resource Injection Enabled
- Tag Pooling Disabled  +  Resource Injection Enabled
- Tag Pooling Enabled  +  Resource Injection Disabled

This fix creates 3 three jsp methods to reduce the size of the jsp service method (to avoid hitting the 65k byte limit) 

- `public void _jsp_cleanUpTag(Object tag, java.util.ArrayList tagList)`  <--- Calls predestroy / release ( and removed the tag from the list if the list in not null) 
- `public void _jsp_cleanUpTagArrayList(java.util.ArrayList tagList)`   <--- Cleans up List 
- `public void _jsp_tagPostConstruct(Object tag, java.util.ArrayList tagList,  com.ibm.ws.managedobject.ManagedObject mo)`  <-- calls doPostConstruct, addTagHandlerToCdiMap and adds the tag to the list if it's not null.  **THIS METHOD is UNIQUE to LIBERTY**

### CODE REVIEW

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,com.ibm.ws.jsp_fat,com.ibm.ws.jsp_fat_jsp.2.3.toleration,com.ibm.ws.jsp_fat_lWAS

**Comment by volosied** [comment]
The current proposal may not be ideal since some code could be called twice, such as the release method below: 

The duplication may also hit some users by having their JSP reach the 65K byte limit for methods. 

```
    try {
      
      _jspx_th_c_set_1.setPageContext(pageContext);
      _jspx_th_c_set_1.setParent(null);
      _jspx_th_c_set_1.setVar("obj");
      _jspx_th_c_set_1.setValue(new org.apache.jasper.el.JspValueExpression("/EL30CollectionObjectOperations.jsp(194,3) '${list.stream().toArray()}'",_el_expressionfactory.createValueExpression(new org.apache.jasper.el.ELContextWrapper(pageContext.getELContext(),_jspx_fnmap),"${list.stream().toArray()}",java.lang.Object.class)).getValue(pageContext.getELContext()));
      _jspx_th_c_set_1.setScope("request");
      int _jspx_eval_c_set_1 = _jspx_th_c_set_1.doStartTag();
      if (_jspx_th_c_set_1.doEndTag() == javax.servlet.jsp.tagext.Tag.SKIP_PAGE) {
        
        _jspx_th_c_set_1.release();
        return true;
      }
     } catch(javax.servlet.jsp.JspException e) {
          throw e;
    } finally {
      
      _jspx_th_c_set_1.release();
    }
```

**Comment by volosied** [comment]
looks better 

But still need to review with team and verify the functionality is same as before. 
```
  private boolean _jspx_meth_sample_Hello_0(PageContext pageContext)
     throws Throwable {
    JspWriter out = pageContext.getOut();
    com.ibm.ws.managedobject.ManagedObject _jspx_th_sample_Hello_0_mo = _jspx_iaHelper.inject(test.HelloWorldTag.class);
    test.HelloWorldTag _jspx_th_sample_Hello_0 = (test.HelloWorldTag)_jspx_th_sample_Hello_0_mo.getObject();
    _jspx_iaHelper.doPostConstruct(_jspx_th_sample_Hello_0);
    _jspx_iaHelper.addTagHandlerToCdiMap(_jspx_th_sample_Hello_0, _jspx_th_sample_Hello_0_mo);
    try {
      
      _jspx_th_sample_Hello_0.setPageContext(pageContext);
      _jspx_th_sample_Hello_0.setParent(null);
      int _jspx_eval_sample_Hello_0 = _jspx_th_sample_Hello_0.doStartTag();
      if (_jspx_eval_sample_Hello_0 != javax.servlet.jsp.tagext.Tag.SKIP_BODY) {
        out = pageContext.getOut();
        out.write(_jsp_string3);
      }
      if (_jspx_th_sample_Hello_0.doEndTag() == javax.servlet.jsp.tagext.Tag.SKIP_PAGE) {
        return true;
      }
     } catch(javax.servlet.jsp.JspException e) {
          throw e;
    } finally {
      _jspx_iaHelper.doPreDestroy(_jspx_th_sample_Hello_0);
      _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_sample_Hello_0);
      _jspx_th_sample_Hello_0.release();
    }
    return false;
  }
  ```

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,com.ibm.ws.jsp_fat,com.ibm.ws.jsp_fat_jsp.2.3.toleration,com.ibm.ws.jsp_fat_lWAS

**Comment by volosied** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer_fat_servlet31,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.1,com.ibm.ws.jsp_fat,com.ibm.ws.jsp.2.3_fat,com.ibm.ws.jsp_fat_lWAS,com.ibm.ws.jsp_fat_jsp.2.3.toleration,com.ibm.ws.el.3.0_fat

**Comment by volosied** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer_fat_servlet31,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.1,com.ibm.ws.jsp_fat,com.ibm.ws.jsp.2.3_fat,com.ibm.ws.jsp_fat_lWAS,com.ibm.ws.jsp_fat_jsp.2.3.toleration,com.ibm.ws.el.3.0_fat

**Comment by volosied** [comment]
Removed catch block since it was necessary.
```
  private boolean _jspx_meth_decorator_title_1(PageContext pageContext)
     throws Throwable {
    JspWriter out = pageContext.getOut();
    com.ibm.ws.managedobject.ManagedObject _jspx_th_decorator_title_1_mo = _jspx_iaHelper.inject(com.opensymphony.module.sitemesh.taglib.decorator.TitleTag.class);
    com.opensymphony.module.sitemesh.taglib.decorator.TitleTag _jspx_th_decorator_title_1 = (com.opensymphony.module.sitemesh.taglib.decorator.TitleTag)_jspx_th_decorator_title_1_mo.getObject();
    _jspx_iaHelper.doPostConstruct(_jspx_th_decorator_title_1);
    _jspx_iaHelper.addTagHandlerToCdiMap(_jspx_th_decorator_title_1, _jspx_th_decorator_title_1_mo);
    try {
      
      _jspx_th_decorator_title_1.setPageContext(pageContext);
      _jspx_th_decorator_title_1.setParent(null);
      int _jspx_eval_decorator_title_1 = _jspx_th_decorator_title_1.doStartTag();
      if (_jspx_th_decorator_title_1.doEndTag() == javax.servlet.jsp.tagext.Tag.SKIP_PAGE) {
        return true;
      }
    } finally {
      _jspx_iaHelper.doPreDestroy(_jspx_th_decorator_title_1);
      _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_decorator_title_1);
      
      _jspx_th_decorator_title_1.release();
    }
    return false;
  }
  ```

**Comment by volosied** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.jsp_fat,com.ibm.ws.jsp.2.3_fat,com.ibm.ws.jsp_fat_lWAS,com.ibm.ws.jsp_fat_jsp.2.3.toleration,com.ibm.ws.el.3.0_fat

**Comment by epj** [comment]
The difference in the generated java from my LeakyRedirect.war test.   This looks good to me.  The finally{} blocks all have the destroy/clean/releases needed.

```140c140
<       
---
>       try {
154c154
<           
---
>             try {
172,175d171
<               _jspx_iaHelper.doPreDestroy(_jspx_th_c_catch_0);
<               _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_c_catch_0);
<               
<               _jspx_th_c_catch_0.release();
185a182
>             } finally {
188d184
<           
189a186
>             }
197,200d193
<         _jspx_iaHelper.doPreDestroy(_jspx_th_c_if_5);
<         _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_c_if_5);
<         
<         _jspx_th_c_if_5.release();
203a197
>       } finally {
206d199
<       
207a201
>       }
245c239
<     
---
>     try {
252,255d245
<       _jspx_iaHelper.doPreDestroy(_jspx_th_c_set_0);
<       _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_c_set_0);
<       
<       _jspx_th_c_set_0.release();
257a248
>     } finally {
260d250
<     
261a252
>     }
273c264
<     
---
>     try {
279,282d269
<       _jspx_iaHelper.doPreDestroy(_jspx_th_c_redirect_0);
<       _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_c_redirect_0);
<       
<       _jspx_th_c_redirect_0.release();
284a272
>     } finally {
287d274
<     
288a276
>     }
300c288
<     
---
>     try {
317,320d304
<       _jspx_iaHelper.doPreDestroy(_jspx_th_c_if_0);
<       _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_c_if_0);
<       
<       _jspx_th_c_if_0.release();
322a307
>     } finally {
325d309
<     
326a311
>     }
338c323
<     
---
>     try {
344,347d328
<       _jspx_iaHelper.doPreDestroy(_jspx_th_c_redirect_1);
<       _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_c_redirect_1);
<       
<       _jspx_th_c_redirect_1.release();
349a331
>     } finally {
352d333
<     
353a335
>     }
365c347
<     
---
>     try {
382,385d363
<       _jspx_iaHelper.doPreDestroy(_jspx_th_c_if_1);
<       _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_c_if_1);
<       
<       _jspx_th_c_if_1.release();
387a366
>     } finally {
390d368
<     
391a370
>     }
403c382
<     
---
>     try {
409,412d387
<       _jspx_iaHelper.doPreDestroy(_jspx_th_c_out_0);
<       _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_c_out_0);
<       
<       _jspx_th_c_out_0.release();
414a390
>     } finally {
417d392
<     
418a394
>     }
430c406
<     
---
>     try {
447,450d422
<       _jspx_iaHelper.doPreDestroy(_jspx_th_c_if_2);
<       _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_c_if_2);
<       
<       _jspx_th_c_if_2.release();
452a425
>     } finally {
455d427
<     
456a429
>     }
468c441
<     
---
>     try {
475,478d447
<       _jspx_iaHelper.doPreDestroy(_jspx_th_c_set_1);
<       _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_c_set_1);
<       
<       _jspx_th_c_set_1.release();
480a450
>     } finally {
483d452
<     
484a454
>     }
496c466
<     
---
>     try {
513,516d482
<       _jspx_iaHelper.doPreDestroy(_jspx_th_c_if_3);
<       _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_c_if_3);
<       
<       _jspx_th_c_if_3.release();
518a485
>     } finally {
521d487
<     
522a489
>     }
534c501
<     
---
>     try {
540,543d506
<       _jspx_iaHelper.doPreDestroy(_jspx_th_c_remove_0);
<       _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_c_remove_0);
<       
<       _jspx_th_c_remove_0.release();
545a509
>     } finally {
548d511
<     
549a513
>     }
561c525
<     
---
>     try {
578,581d541
<       _jspx_iaHelper.doPreDestroy(_jspx_th_c_if_4);
<       _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_c_if_4);
<       
<       _jspx_th_c_if_4.release();
583a544
>     } finally {
586d546
<     
587a548
>     }
599c560
<     
---
>     try {
613,616d573
<       _jspx_iaHelper.doPreDestroy(_jspx_th_c_when_0);
<       _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_c_when_0);
<       
<       _jspx_th_c_when_0.release();
618a576
>     } finally {
621d578
<     
622a580
>     }
634c592
<     
---
>     try {
647,650d604
<       _jspx_iaHelper.doPreDestroy(_jspx_th_c_otherwise_0);
<       _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_c_otherwise_0);
<       
<       _jspx_th_c_otherwise_0.release();
652a607
>     } finally {
655d609
<     
656a611
>     }
668c623
<     
---
>     try {
687,690d641
<       _jspx_iaHelper.doPreDestroy(_jspx_th_c_choose_0);
<       _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_c_choose_0);
<       
<       _jspx_th_c_choose_0.release();
692a644
>     } finally {
695d646
<     
696a648
>     }
708c660
<     
---
>     try {
725,728d676
<       _jspx_iaHelper.doPreDestroy(_jspx_th_c_if_6);
<       _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_c_if_6);
<       
<       _jspx_th_c_if_6.release();
730a679
>     } finally {
733d681
<     
734a683
>     }
746c695
<     
---
>     try {
755,758d703
<         _jspx_iaHelper.doPreDestroy(_jspx_th_c_import_0);
<         _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_c_import_0);
<         
<         _jspx_th_c_import_0.release();
767a713
>     } finally {
770d715
<     
771a717
>     }
783c729
<     
---
>     try {
800,803d745
<       _jspx_iaHelper.doPreDestroy(_jspx_th_c_if_7);
<       _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_c_if_7);
<       
<       _jspx_th_c_if_7.release();
805a748
>     } finally {
808d750
<     
809a752
>     }
821c764
<     
---
>     try {
828,831d770
<       _jspx_iaHelper.doPreDestroy(_jspx_th_c_param_0);
<       _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_c_param_0);
<       
<       _jspx_th_c_param_0.release();
833a773
>     } finally {
836d775
<     
837a777
>     }
849c789
<     
---
>     try {
872,875d811
<       _jspx_iaHelper.doPreDestroy(_jspx_th_c_url_0);
<       _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_c_url_0);
<       
<       _jspx_th_c_url_0.release();
877a814
>     } finally {
880d816
<     
881a818
>     }
893c830
<     
---
>     try {
910,913d846
<       _jspx_iaHelper.doPreDestroy(_jspx_th_c_if_8);
<       _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_c_if_8);
<       
<       _jspx_th_c_if_8.release();
915a849
>     } finally {
918d851
<     
919a853
>     }
931c865
<     
---
>     try {
937,940d870
<       _jspx_iaHelper.doPreDestroy(_jspx_th_c_out_1);
<       _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_c_out_1);
<       
<       _jspx_th_c_out_1.release();
942a873
>     } finally {
945d875
<     
946a877
>     }
958c889
<     
---
>     try {
979,982d909
<         _jspx_iaHelper.doPreDestroy(_jspx_th_c_forTokens_0);
<         _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_c_forTokens_0);
<         
<         _jspx_th_c_forTokens_0.release();
991a919
>     } finally {
994d921
<     
995a923
>     }
1007c935
<     
---
>     try {
1024,1027d951
<       _jspx_iaHelper.doPreDestroy(_jspx_th_c_if_9);
<       _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_c_if_9);
<       
<       _jspx_th_c_if_9.release();
1029a954
>     } finally {
1032d956
<     
1033a958
>     }
1039c964
< /opt/Liberty20009/wlp/usr/servers/server1/workarea/org.eclipse.osgi/73/data/temp/default_node/SMF_WebContainer/LeakyRedirect/LeakyRedirect/_index.java was generated @ Mon Sep 19 12:34:06 PDT 2022
---
> /opt/Liberty20009/wlp/usr/servers/server1/workarea/org.eclipse.osgi/73/data/temp/default_node/SMF_WebContainer/LeakyRedirect/LeakyRedirect/_index.java was generated @ Mon Sep 19 12:26:02 PDT 2022
```

**Comment by volosied** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.jsp_fat,com.ibm.ws.jsp.2.3_fat,com.ibm.ws.jsp_fat_lWAS,com.ibm.ws.jsp_fat_jsp.2.3.toleration,com.ibm.ws.el.3.0_fat

**Comment by volosied** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.jsp_fat,com.ibm.ws.jsp.2.3_fat,com.ibm.ws.jsp_fat_lWAS,com.ibm.ws.jsp_fat_jsp.2.3.toleration,com.ibm.ws.el.3.0_fat

**Review by epj — DISMISSED** [review]
I'm happy to see this finally resolved.  The changes look good and customer had success with the test fix.

**Comment by volosied** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer_fat_servlet31,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.1,com.ibm.ws.jsp_fat,com.ibm.ws.jsp.2.3_fat,com.ibm.ws.jsp_fat_lWAS,com.ibm.ws.jsp_fat_jsp.2.3.toleration,com.ibm.ws.el.3.0_fat

**Comment by volosied** [comment]
Users are reporting this change is hitting the 65535 bytes limit for the _jspService method.

Changed to a draft as I work on a fix.

**Comment by volosied** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.jsp_fat,com.ibm.ws.jsp.2.3_fat,com.ibm.ws.jsp_fat_lWAS,com.ibm.ws.jsp_fat_jsp.2.3.toleration,com.ibm.ws.el.3.0_fat

**Comment by volosied** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.jsp_fat,com.ibm.ws.jsp.2.3_fat,com.ibm.ws.jsp_fat_lWAS,com.ibm.ws.jsp_fat_jsp.2.3.toleration,com.ibm.ws.el.3.0_fat

**Comment by volosied** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.jsp_fat,com.ibm.ws.jsp.2.3_fat,com.ibm.ws.jsp_fat_lWAS,com.ibm.ws.jsp_fat_jsp.2.3.toleration,com.ibm.ws.el.3.0_fat

**Comment by volosied** [comment]
I think this is the way the methods should be generated: 
```
  private boolean _jspx_meth_sample_Hello_0(PageContext pageContext)
  throws Throwable {
 com.ibm.ws.managedobject.ManagedObject _jspx_th_sample_Hello_0_mo = null;
 try {
    JspWriter out = pageContext.getOut();
    jspx_th_sample_Hello_0_mo = _jspx_iaHelper.inject(test.HelloWorldTag.class);
    test.HelloWorldTag _jspx_th_sample_Hello_0 = (test.HelloWorldTag)_jspx_th_sample_Hello_0_mo.getObject();
    _jspx_iaHelper.doPostConstruct(_jspx_th_sample_Hello_0);
    _jspx_iaHelper.addTagHandlerToCdiMap(_jspx_th_sample_Hello_0, _jspx_th_sample_Hello_0_mo);
    _jspx_th_sample_Hello_0.setPageContext(pageContext);
    _jspx_th_sample_Hello_0.setParent(null);
    int _jspx_eval_sample_Hello_0 = _jspx_th_sample_Hello_0.doStartTag();
    if (_jspx_eval_sample_Hello_0 != javax.servlet.jsp.tagext.Tag.SKIP_BODY) {
      out = pageContext.getOut();
      out.write(_jsp_string3);
    }
    if (_jspx_th_sample_Hello_0.doEndTag() == javax.servlet.jsp.tagext.Tag.SKIP_PAGE) { <<----- EXCEPTION THROWN HERE
      return true;
    }
 } finally {
  if(_jspx_th_sample_Hello_0_mo != null){
    _jspx_iaHelper.doPreDestroy(_jspx_th_sample_Hello_0);
    _jspx_iaHelper.cleanUpTagHandlerFromCdiMap(_jspx_th_sample_Hello_0);
    _jspx_th_sample_Hello_0.release();
  }
 }
 return false;
}
```

File: dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/translator/visitor/generator/CustomTagGenerator.java [hunk 1]
```diff
@@ -560,45 +492,51 @@ private void assembleTagFromWriters(JavaCodeWriter writer) throws JspCoreExcepti
...
 
         if (genTagInMethod) {
             if (methodNesting > 0) {
                 methodWriter.println("return false;");
+
+                if ((reuseTag == false) && !(tagGenerator instanceof SimpleTagGenerator)) {
+                    methodWriter.println("} finally {");
+                    methodWriter.println("_jsp_cleanUpTag(" + tagHandlerVar + ", null);");
+                    methodWriter.println("}");
+                }
```
Comment (volosied) [inline dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/translator/visitor/generator/CustomTagGenerator.java:509]: Previously, the clean up code (i.e doPreDestroy) was in the generateEndTag methods. I've placed them it here in this finally block again.

This only applies if we are within a tag method.

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,com.ibm.ws.jsp_fat,com.ibm.ws.jsp_fat_jsp.2.3.toleration,com.ibm.ws.jsp_fat_lWAS

**Comment by volosied** [comment]
#build

File: dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPCdiTest.java [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2013, 2020 IBM Corporation and others.
+ * Copyright (c) 2013, 2022 IBM Corporation and others.
```
Comment (pnicolucci) [inline dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPCdiTest.java:2]: Is 2022 correct here? Code first pushed up in 2022?

File: dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/inmemory/generator/InMemoryGenerateJspVisitor.java [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 1997, 2005 IBM Corporation and others.
+ * Copyright (c) 1997, 2023 IBM Corporation and others.
```
Comment (pnicolucci) [inline dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/inmemory/generator/InMemoryGenerateJspVisitor.java:2]: Can we also remove the contributors?

File: dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/inmemory/generator/InMemoryGenerateTagFileVisitor.java [hunk 4]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 1997, 2005 IBM Corporation and others.
+ * Copyright (c) 1997, 2023 IBM Corporation and others.
```
Comment (pnicolucci) [inline dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/inmemory/generator/InMemoryGenerateTagFileVisitor.java:2]: Can we also remove the contributors?

File: dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/translator/visitor/generator/CustomTagGenerator.java [hunk 5]
```diff
@@ -44,75 +44,73 @@
...
                      ValidateResult validatorResult,
                      JspVisitorInputMap inputMap,
                      ArrayList methodWriterList,
-                     FragmentHelperClassWriter fragmentHelperClassWriter, 
+                     FragmentHelperClassWriter fragmentHelperClassWriter,
                      HashMap persistentData,
                      JspConfiguration jspConfiguration,
-                     JspOptions jspOptions) 
-        throws JspCoreException {
+                     JspOptions jspOptions) throws JspCoreException {
```
Comment (pnicolucci) [inline dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/translator/visitor/generator/CustomTagGenerator.java:95]: Copyright date range update is missing

File: dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/translator/visitor/generator/FragmentHelperClassWriter.java [hunk 6]
```diff
@@ -17,6 +17,7 @@
 import org.w3c.dom.Element;
 
 import com.ibm.ws.jsp.JspCoreException;
+import com.ibm.ws.jsp.JspOptions;
```
Comment (pnicolucci) [inline dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/translator/visitor/generator/FragmentHelperClassWriter.java:17]: Copyright date range update is missing

File: dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPCdiTest.java (hunk 2 above)
Comment (volosied) [reply dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPCdiTest.java:2]: correct

**Review by pnicolucci — APPROVED** [review]

**Comment by volosied** [comment]
#libby

**Pull request merged**